enhancements/windows-containers/windows-machine-config-operator-upgrades.md

```diff
@@ -0,0 +1,188 @@
+the terminated ones.
+
+In order to allow for this to occur with minimal service disruption, it is
+recommended that users have at least two Windows nodes within the cluster.
+
+All Windows Kubernetes component updates will be tied to WMCO releases.
+
+### Justification
+
+The reason we need to cause the VMs to be fully terminated, instead of updating
+them, is that changes made to the VM done in previous versions of the WMCO may
+cause issues with the current version. If we chose to do an in place update,
+we would need to actively undo changes made to the node that may cause issues
+with the current WMCO version. The logic to keep track and undo these changes
+adds an additional layer of complexity that also adds the benefit of a quicker
+upgrade time. The risk of introducing additional issues is not worth it at this
+time, however it should be explored as the product matures.
+
+The reason we will not handle Windows operating system updates is that the
+cluster administrator provides the Windows image to create the VMs with, and
+it stays their responsiblity to provide an updated image.
+
+It is left to the user to create the Windows MachineHealthChecks, as
+WMCO's job is solely to configure Windows VMs into nodes. It is the user's
+responsibility to create the MachineSet which in turn creates the Windows VMs.
+Because of that, the responsibility falls onto the user to set up
+MachineHealthChecks, as WMCO does not create the Machines, and should not
+decide when they should be destroyed.
+
+### Design Details
+
+WMCO is published to OperatorHub as a Red Hat operator. When a new WMCO version
+is released, and the new version is compatible with the current cluster
+version, an update can occur. While we cannot explicity specify a required OCP
+version for the WMCO upgrade, each new OCP version upgrade should include a
+Kubernetes version upgrade. This allows us to use the [minKubeVersion](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/building-your-csv.md#operator-metadata)
+field in WMCO's CSV, which stops the operator from upgrading if the next
+WMCO version does not have at least the specified Kubernetes version.
+
+This update will either happen automatically, or require approval, based on
+the settings given through OLM when initially installing the operator.
+
+In order to facilitate upgrades, WMCO will add an annotation to the node
+indicating the version of the operator which configured the node.
+Using a node annotation is the easiest way to track this information in a way
+that lasts across WMCO restarts. WMCO is directly responsible for the creation
+of Windows nodes, and we should expect that other entities will not clear or
+alter the annotations WMCO gives to such nodes.
+
+On reconcile, if the annotated version does not match the current version,
+or the annotation is unexpectedly missing, WMCO will send a stop signal to the
```

Comment:
This sort of seems to a side-effect right. For example, you can end-up in the scenario when the upgrade is not needed but annotation is present. However since WMCB cannot distinguish between an upgrade and missing node-annotation, I am fine as long as we log appropriately or generate events.

Reply:
Yes I think we should be logging the reason for termination when we bring down a node, whether its for upgrade, normal scale down or missing annotation